src/client.ts

```diff
@@ -317,6 +317,35 @@ in an environmental variable named FAUNA_SECRET or pass it to the Client\
     return maybeSecret;
   }
 
+  #getEndpoint(partialClientConfig?: ClientConfiguration): URL {
+    // If the user explicitly sets the endpoint to undefined, we should throw a
+    // TypeError, rather than override with the default endpoint.
+    if (
+      partialClientConfig &&
+      "endpoint" in partialClientConfig &&
+      partialClientConfig.endpoint === undefined
+    ) {
+      throw new TypeError(
+        `ClientConfiguration option endpoint must be defined.`
+      );
+    }
```

Comment:
I actually think we shouldn't care if someone sets it to null or undefined. If they do, let it coalesce to the env var or the default value.

If you do think this is important, why not disallow null as well?

Reply:
We've been here before 😄 I ported over an `applyDefaults` function from the v4 driver that treated explicit undefined values the same as not-provided, thus applying defaults on top of them. 

https://github.com/fauna/fauna-js/pull/145#discussion_r1187792805

And in response: 

> Treating undefined special makes the interface more complex

https://github.com/fauna/fauna-js/pull/145#discussion_r1188797204

What we get by not accepting explicitly set undefined values is we catch what are most likely (if, admittedly, not always) mistakes/bugs in user code.

Regarding `null`: there are plenty of other bad inputs one could provide, including `null`, but we took the path of focusing on enforcing through TS types. Our types are actually currently enforcing that provided values must be defined (we specify `field?: Type` and not `field?: Type | undefined`), as well as not `null`. But it's a really common JS bug that assumes a variable is defined but isn't. It it more rare to accidentally assign a variable to null when it should have a value.

All that said, I do not mean to argue very strongly for keeping the errors for explicit undefined -- I'm really just trying to lay out the history for the record. I think this is a design choice where neither option is more "correct".

It would technically be a breaking change to start overriding explicit undefined values. The types disallow explicit undefined, so what we would be doing is changing an explicit error to one we handle internally for the user by applying our own defaults.

Reply:
Okay, if we've been here we've been here :) I'm just in favor of less code, but I'll ship this.

If it comes up again, hopefully I'll remember.